Added MIP for fix error handling in API Responses (#36)
MetaMask/metamask-improvement-proposals: MIPs/mip-x .md

---
MIP: X
Title: Fix Error Handling in API Responses
Status: Draft
Stability: Experimental

> discussions-to: https://github.com/MetaMask/metamask-improvement-proposals/discussions

Author: ```suggestion
discussions-to: https://github.com/MetaMask/metamask-improvement-proposals/discussions/37
```

---
MIP: X
Title: Fix Error Handling in API Responses
Status: Draft
Stability: Experimental

> discussions-to: https://github.com/MetaMask/metamask-improvement-proposals/discussions

Reviewer: reminder to update this

Author(s): Shane Jonas (@shanejonas)
Type: Maintaner
Created: 2024-03-04
---

#### Abstract

> This proposal aims to refine MetaMask's API error handling by eliminating the practice of stringifying or wrapping errors. The current method obscures underlying issues, complicating debugging and resolution for developers.

Reviewer: What do you think about using some other word to describe the process/pipeline instead of `method`?

#### Motivation

> Issue https://github.com/MetaMask/metamask-extension/issues/19697 and https://github.com/MetaMask/metamask-extension/pull/15205#issuecomment-1199953855 highlighted difficulties faced by developers due to the API's error handling mechanism, where errors were stringified/wrapped, masking the root cause. This change proposes to break some existing error cases to facilitate clearer understanding and quicker resolution of issues.

Reviewer: the errors are both wrapped and stringified, right? would it be more accurate to say that?

#### Specification
- **Current Behaviour:** Errors returned by the API are wrapped and stringified, leading to terribly formatted error messages in some cases.
- **Proposed Change:** Modify the error handling process to return original error messages directly without wrapping or stringification. This would introduce a breaking change to developers who have wrote code to parse those errors and handle them.

#### Example
There are some edge cases with sendTransaction where a node may have formatted the error message in a way that is not expected.

```javascript
window.ethereum.request({ method: 'eth_sendTransaction', params: [tx] })
  .catch((error) => {
    console.error(error);
  });
```

For example this one is from ganache and is wrapped in `value`.
##### Current Behaviour
```

> // Logs:

Reviewer: ```suggestion
```

{
  "jsonrpc":"2.0",
  "error":"[ethjs-query] while formatting outputs from RPC '{"value":{"code":-32603}}'",
  "id": 4653223632683671
}
```

##### Proposed Behaviour

> ```javascript

Reviewer: ```suggestion
```json
```

---
MIP: X
Title: Fix Error Handling in API Responses
Status: Draft
Stability: Experimental
discussions-to: https://github.com/MetaMask/metamask-improvement-proposals/discussions/37
Author(s): Shane Jonas (@shanejonas)
Type: Maintaner
Created: 2024-03-04
---

#### Abstract
This proposal aims to refine MetaMask's API error handling by eliminating the practice of stringifying or wrapping errors. The current edge cases obscure underlying issues, complicating debugging and resolution for developers.

#### Motivation
Issue https://github.com/MetaMask/metamask-extension/issues/19697, https://github.com/MetaMask/metamask-extension/issues/15250 and https://github.com/MetaMask/metamask-extension/pull/15205#issuecomment-1199953855 highlighted difficulties faced by developers due to the API's error handling mechanism, where some errors are wrapped and stringified. This change proposes to break some existing error cases to facilitate clearer understanding and quicker resolution of issues.

#### Specification

> - **Current Behaviour:** Errors returned by the API are wrapped and stringified, leading to terribly formatted error messages in some cases.

Reviewer 2: Hmm, I had always thought of this as a bug, rather than a part of how the API works. But it's a fair point that we would expect dapps to expect these responses.

Do we know if we return these mangled stringified errors for all errors? Or just certain classes of them? I thought we returned proper Ethereum JSON-RPC errors at least some of the time.

Reviewer 3: Would be good to drill in on what specific methods/scenarios we see these poorly formatted errors.

---
MIP: X
Title: Fix Error Handling in API Responses
Status: Draft

> Stability: Experimental

Reviewer 4: ```suggestion
Stability: n/a
```
We typically keep the status as n/a until implemented

discussions-to: https://github.com/MetaMask/metamask-improvement-proposals/discussions/37
Author(s): Shane Jonas (@shanejonas)

> Type: Maintaner

Reviewer 3: ```suggestion
Type: Maintainer
```

#### Specification
- **Current Behaviour:** Errors returned by the API are wrapped and stringified, leading to terribly formatted error messages in some cases.
- **Proposed Change:** Modify the error handling process to return original error messages directly without wrapping or stringification. This would introduce a breaking change to developers who have wrote code to parse those errors and handle them.

#### Examples
There are some edge cases with sendTransaction where a node may have formatted the error message in a way that is not expected.

```javascript
window.ethereum.request({ method: 'eth_sendTransaction', params: [tx] })
  .catch((error) => {
    console.error(error);
  });
```

##### Current Behaviour
```json
{
  "jsonrpc": "2.0",
  "error": "[ethjs-query] while formatting outputs from RPC '{\"value\":{\"code\":-32603,\"message\":\"Internal Server Error\"}}'",
  "id": 4653223632683671
}
```
```json
{
  "jsonrpc": "2.0",
  "error": "[ethjs-query] while formatting outputs from RPC '{\"value\":{\"code\":-32603,\"data\":{\"message\":\"Internal Server Error\"}}}'",
}
```
```json
{
  "id": 0,
  "error": {
    "code": -32603,

>     "message": "[ethjs-query] while formatting outputs from RPC '{\"value\":{\"code\":-32603,\"data\":{\"message\":\"VM Exception while processing transaction: invalid opcode\",\"code\":-32000,\"data\":{\"0x877c456d0b406f0706f405b0d2eb4e0ebc69629a539503e7ad94a7f9b6ba36af\":{\"error\":\"invalid opcode\",\"program_counter\":13,\"return\":\"0x\"},\"stack\":\"c: VM Exception while processing transaction: invalid opcode\\n    at Function.c.fromResults (/home/x/code/x/smart-contracts/node_modules/ganache-cli/build/ganache-core.node.cli.js:4:192416)\\n    at w.processBlock (/home/x/code/x/smart-contracts/node_modules/ganache-cli/build/ganache-core.node.cli.js:42:50915)\\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\",\"name\":\"c\"}}}}'"

Reviewer 5: what is the proposed behavior if the returned message including those information ? Looks current metamask only return `Internal server error`

  }
}
```

##### Proposed Behaviour
```json
{ 
  "jsonrpc":"2.0",
  "error": {
    "message": "Internal Server Error",
    "code": -32603,
  },
  "id": 4653223632683671
}
```

#### Rationale
Directly presenting errors enhances developer experience by providing clear, actionable insights into issues encountered during development, thereby reducing troubleshooting time and effort. This change would also align MetaMask's error handling with the broader Ethereum ecosystem, making it easier for developers to work with MetaMask.

#### Backwards Compatibility

> This change is expected to **NOT** be backwards compatible, as it involves changing the error message formatting for some cases. Developers accustomed to the existing error handling mechanism for those cases may need to adjust their code.

Reviewer 4: ```suggestion
This change is **NOT** expected to be backwards compatible, as it involves changing the error message formatting for some cases. This change corrects unintended behaviors to follow the intended design more closely. However, developers expecting the existing error handling mechanism and formats may need to adjust their code.
```
